File: src/main/kotlin/com/ivankresic/kotlinadvanced/classes/NestedClasses.java
package com.ivankresic.kotlinadvanced.classes;

/**
 * Created by ivan on 04/06/2020
 */
public class NestedClasses {
    public static void main(String[] args){
        //DirectoryExplorer.PermissionCheck permissionCheck = new DirectoryExplorer.PermissionCheck();


        //In order to enable this we have to annotate companion object in Kotlin with "@JvmStatic"
        Log log =  Log.createFileLog("");
    }
}

File: src/main/kotlin/com/ivankresic/kotlinadvanced/classes/NestedClasses.kt
package com.ivankresic.kotlinadvanced.classes

/**
 * Created by ivan on 04/06/2020
 */
class DirectoryExplorer(val user:String){

    //we can use private to hide it from upper usage

    /**
     * Defining class as inner class, we can access properties of outer class
     * */
    inner class PermissionCheck(){
        fun validatePassword() {
            if(user == "Ivan")
                println("Welcome $user")
        }

    }

    fun listFolder(folfer: String, user: String){
        val permissionCheck = PermissionCheck()

        permissionCheck.validatePassword()
    }
}

fun main(args: Array<String>) {
    val de = DirectoryExplorer("Ivan")
    //val pc = DirectoryExplorer.PermissionCheck()  ->  before inner

    //now it is accessible as instance of a class
    val pc = DirectoryExplorer("Ivan").PermissionCheck() // ->  after inner
}